tsdb/index/index.go

```diff
@@ -993,39 +972,126 @@ func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
+type Symbols struct {
+	bs      ByteSlice
+	version int
+	off     int
+
+	index []int
+	seen  int
+}
+
+const SYMBOL_FACTOR = 32
+
+func NewSymbols(bs ByteSlice, version int, off int) (*Symbols, error) {
+	s := &Symbols{
+		bs:      bs,
+		version: version,
+		off:     off,
+	}
 	if off == 0 {
-		return nil, nil, nil
+		// Only happens in some tests.
+		return nil, nil
 	}
 	d := encoding.NewDecbufAt(bs, off, castagnoliTable)
-
 	var (
-		origLen     = d.Len()
-		cnt         = d.Be32int()
-		basePos     = uint32(off) + 4
-		nextPos     = basePos + uint32(origLen-d.Len())
-		symbolSlice []string
-		symbols     = map[uint32]string{}
+		origLen = d.Len()
+		cnt     = d.Be32int()
+		basePos = off + 4
 	)
-	if version == FormatV2 {
-		symbolSlice = make([]string, 0, cnt)
+	s.index = make([]int, 0, cnt/SYMBOL_FACTOR)
+	for d.Err() == nil && s.seen < cnt {
+		if s.seen%SYMBOL_FACTOR == 0 {
+			s.index = append(s.index, basePos+origLen-d.Len())
+		}
+		d.UvarintBytes() // The symbol.
+		s.seen++
 	}
+	if d.Err() != nil {
+		return nil, d.Err()
+	}
+	return s, nil
+}
 
-	for d.Err() == nil && d.Len() > 0 && cnt > 0 {
-		s := d.UvarintStr()
+func (s Symbols) Lookup(o uint32) (string, error) {
+	d := encoding.Decbuf{
+		B: s.bs.Range(0, s.bs.Len()),
+	}
+	if s.version == FormatV2 {
+		if int(o) > s.seen {
+			return "", errors.Errorf("unknown symbol offset %d", o)
+		}
+		d.Skip(s.index[int(o/SYMBOL_FACTOR)])
+		// Walk until we find the one we want.
+		for i := o - (o / SYMBOL_FACTOR * SYMBOL_FACTOR); i > 0; i-- {
+			d.UvarintBytes()
+		}
+	} else {
+		d.Skip(int(o))
+	}
+	sym := d.UvarintStr()
+	if d.Err() != nil {
+		return "", d.Err()
+	}
+	return sym, nil
+}
 
-		if version == FormatV2 {
-			symbolSlice = append(symbolSlice, s)
-		} else {
-			symbols[nextPos] = s
-			nextPos = basePos + uint32(origLen-d.Len())
+func (s Symbols) ReverseLookup(sym string) (uint32, error) {
+	i := sort.Search(len(s.index), func(i int) bool {
+		// Any decoding errors here will be lost, however
+		// we already read through all of this at startup.
+		d := encoding.Decbuf{
+			B: s.bs.Range(0, s.bs.Len()),
+		}
+		d.Skip(s.index[i])
+		return yoloString(d.UvarintBytes()) > sym
+	})
+	d := encoding.Decbuf{
+		B: s.bs.Range(0, s.bs.Len()),
+	}
+	if i > 0 {
+		i--
+	}
+	d.Skip(s.index[i])
+	res := i * 32
+	var lastLen int
+	var lastSymbol string
+	for d.Err() == nil && res <= s.seen {
+		lastLen = d.Len()
+		lastSymbol = yoloString(d.UvarintBytes())
+		if lastSymbol >= sym {
+			break
 		}
+		res++
+	}
+	if d.Err() != nil {
+		return 0, d.Err()
+	}
+	if lastSymbol != sym {
+		return 0, errors.Errorf("unknown symbol %q", sym)
+	}
+	if s.version == FormatV2 {
+		return uint32(res), nil
+	}
+	return uint32(s.bs.Len() - lastLen), nil
+}
+
+func (s Symbols) All() (map[string]struct{}, error) {
+	d := encoding.NewDecbufAt(s.bs, s.off, castagnoliTable)
+	cnt := d.Be32int()
+	res := make(map[string]struct{}, cnt)
+	for d.Err() == nil && cnt > 0 {
+		res[d.UvarintStr()] = struct{}{}
```

Comment:
why not using yoloStrings here if we use this method only for `populateAll` :thinking:

Reply:
Ok we actually change logic here: https://github.com/prometheus/prometheus/pull/6468 makes sense